src/misc/umap.c

```diff
@@ -0,0 +1,846 @@
+ * in PCA).
+ *
+ * However, UMAP uses a graph with distances associated to the edges as a key
+ * intermediate representation of the high-dimensional space, so it is also useful as
+ * a general graph layouting algorithm, hence its inclusion in igraph.
+ *
+ * Importantly, the edge-associated distances are derived from a similarity metric
+ * between the high-dimensional vectors, often Pearson correlation:
+ *
+ * corr(v1, v2) = v1 x v2 / [ sqrt(v1 x v1) * sqrt(v2 x v2) ]
+ *
+ * In this case, the associated distance is usually defined as:
+ *
+ * d(v1, v2) = 1 - corr(v1, v2)
+ *
+ * This implementation can also work with unweighted similarity graphs, in which case
+ * the distance parameter should be a null pointer and all edges beget a similarity
+ * score of 1 (a distance of 0).
+ *
+ * While all similarity graphs are theoretically embeddable, UMAP's stochastic gradient
+ * descent approach really shines when the graph is sparse. In practice, most people
+ * feed a k-nearest neighbor (either computed exactly or approximated) similarity graph
+ * with some additional cutoff to exclude "quasi-neighbors" that lie beyond a certain
+ * distance (e.g. correlation less than 0.2).
+ *
+ * Therefore, if you are trying to use this function to embed high-dimensional vectors,
+ * the steps are:
+ *
+ * 1. Compute a sparse similarity graph (either exact or approximate) from your vectors,
+ *    weighted or unweighted. If unsure, compute a knn.
+ * 2. If you keep the weights, convert them into distances or store them as a "weight"
+ *    edge attribute and use a null pointer for the distances. If using similarity
+ *    weights instead of distances, make sure they do not exceed 1.
+ * 3. Feed the graph (and distances, if you have them) into this function.
+ *
+ * Note: Step 1 above involves deciding if two high-dimensional vectors "look similar"
+ *       which, because of the curse of dimensionality, is in many cases a highly
+ *       subjective and potentially controversial operation: thread with care and at
+ *       your own risk. Two high-dimensional vectors might look similar or extremely
+ *       different depending on the point of view/angle, and there are a lot of
+ *       viewpoints when the dimensionality ramps up.
+ *
+ * </para><para>
+ * References:
+ *
+ * </para><para>
+ * Leland McInnes, John Healy, and James Melville, arXiv:1802.03426v3
+ *
+ * \param graph Pointer to the similarity graph to find a layout for (i.e. to embed). Weights in
+ *   this graph object are currently ignored.
+ * \param distances Pointer to a vector of "distances" between vertices. Similarity graphs for
+ *   UMAP are often originally meant in terms of similarity weights (e.g. correlation between
+ *   high-dimensional vectors) and converted into distances by crude dist := 1 - corr. That is
+ *   fine here too. If this argument is an empty pointer (NULL), connected vertices are assumed
+ *   to be very similar.
```

Comment:
I don't quite understand this description. Is this an edge property? If so, phrase it accordingly (e.g. "edge lengths" instead of "distances between vertices") to make it clear that `distances[i]` belongs to edge `i`.

I'm not sure what "connected vertices are assumed to be very similar" means.